src/state/mod.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 Damir Jelić
+// Copyright 2020 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::path::Path;
+
+pub mod state_store;
+pub use state_store::JsonStore;
+
+use serde::{Deserialize, Serialize};
+
+use crate::base_client::{Client as BaseClient, Token};
+use crate::events::push_rules::Ruleset;
+use crate::identifiers::{RoomId, UserId};
+use crate::models::Room;
+use crate::session::Session;
+use crate::Result;
+#[derive(Debug, Default, PartialEq, Serialize, Deserialize)]
+pub struct ClientState {
+    /// The current client session containing our user id, device id and access
+    /// token.
+    pub session: Option<Session>,
+    /// The current sync token that should be used for the next sync call.
+    pub sync_token: Option<Token>,
+    /// A list of ignored users.
+    pub ignored_users: Vec<UserId>,
+    /// The push ruleset for the logged in user.
+    pub push_ruleset: Option<Ruleset>,
+}
+
+impl ClientState {
+    pub fn from_base_client(client: &BaseClient) -> ClientState {
+        let BaseClient {
+            session,
+            sync_token,
+            ignored_users,
+            push_ruleset,
+            ..
+        } = client;
+        Self {
+            session: session.clone(),
+            sync_token: sync_token.clone(),
+            ignored_users: ignored_users.clone(),
+            push_ruleset: push_ruleset.clone(),
+        }
+    }
+}
+
+/// Abstraction around the data store to avoid unnecessary request on client initialization.
+#[async_trait::async_trait]
+pub trait StateStore: Send + Sync {
+    /// Set up connections or check files exist to load/save state.
+    fn open(&self, path: &Path) -> Result<()>;
```

Comment:
Because of the way we pass a custom store implementation to the client (we pass an already opened/created store) we don't need to make this a method under the trait. It can be just a normal method of the `JsonStore`. That way we don't restrict the arguments and return types for it unnecessarily.